File: FpWorkshop/FSharpPortion/Step02_Functions.fs
namespace FSharpPortion

module Step02_Functions =

  let run () =
    // Challenge 1 --------------------------------------
    let iTellYouWhatYouToldMe x = x

    ensure(iTellYouWhatYouToldMe 42 = 42)

    // Challenge 1 (aside) ------------------------------
    /// The above function definition syntax is actually just syntactic sugar for this
    let iTellYouWhatYouToldMeLowCarb = fun x -> x

    ensure(iTellYouWhatYouToldMeLowCarb 42 = 42)

    // Challenge 2 --------------------------------------

    ensure((fun x -> x) "Wait! You lost your name" = "Wait! You lost your name")

    // Challenge 3 --------------------------------------
    let myAdd a b = a + b

    // Application is just whitespace in F#, no parens necessary unless expressing order of op!
    ensure(myAdd 2 3 = 5)

    // Challenge 4 --------------------------------------
    let specialAdd = myAdd 4 // Arguments can be applied later, no problem!

    ensure(specialAdd 2 = 6)

    // Challenge 5 --------------------------------------
    let alwaysReturnTheFirst a b = a

    // Given this test compiles, how many possible implementations does `alwaysReturnTheFirst` have?
    ensure(alwaysReturnTheFirst "foo" 42 = "foo")

    // Challenge 6 --------------------------------------
    // All that functions really do is bind names to values in a given scope... Sound familiar?
    (
      fun a b c ->
        ensure(b = "blue")
        ensure(a = "red")
        ensure(c = "green")

    ) "red" "blue" "green" 

    ensure(alwaysReturnTheFirst "foo" 42 = "foo")

    // Challenge 6 (aside) ------------------------------
    // It turns out that `let` can be thought of as just syntactic sugar for function application.
    // Can you think of a case where the following 2 forms behave differently?

//    let a = value
//    do some stuff

//    (fun a -> do some stuff) value


    // Challenge 7 --------------------------------------
    // Let's make let, shall we? Hehe

    let sillyLet value scope = scope value

    sillyLet 5 (fun a ->
    sillyLet 6 (fun b ->
      ensure(a + b = 11)
    ))

    // Challenge 8 --------------------------------------
    // Functions can accept functions as arguments
    let applyThenMultiply f a b = f a * f b

    let addOne = myAdd 1

    ensure (applyThenMultiply addOne 4 5 = 30)

    // Challenge 9 --------------------------------------
    // Which function that we've already seen acts like "choose value"?
    ensure (applyThenMultiply iTellYouWhatYouToldMe 7 3 = (7*3))

    // Challenge 9 (spoiler/note) -----------------------
    // This function exists in F#'s global namespace as `id`
    ensure (applyThenMultiply id 7 3 = 21)

    // Challenge 10 --------------------------------------
    // Which function that we've already seen acts like "ignore value, use this instead"?
    ensure (applyThenMultiply (alwaysReturnTheFirst 5) 7 3 = (5*5))

    // Challenge 11 --------------------------------------
    // In F# even symbolic operators are functions!
    // They can be moved to infix notation (apply in the front) by wrapping in parens

    ensure ((*) 3 4 = 3 * 4)

    // Challenge 12 --------------------------------------
    // This is pretty powerful when combined with partial application
    let alwaysAdd21 = (+) 21

    ensure (alwaysAdd21 3 = 24)

    // Challenge 13 --------------------------------------
    // Since functional languages are all about function application
    // sometimes things get a little, well, nested
    let myMult = (*)
    let yuckNesting a = myAdd 1 (myMult 21 (myAdd 4 (myMult 2 a)))

    // So functional languages typically introduce a "pipe" operator so that
    // computation can be thought of as a pipeline

    // In F# this is the `|>` operator. It's defined like so (look familiar?)

    (*
    let (|>) someValue someFunction = someFunction someValue
    *)

    // You can think of this operator as
    // "flowing data through a pipe, left-to-right"

    let pipedResult = 21 |> myAdd 1 |> myMult 2

    ensure (pipedResult = 44)

    // Challenge 14 --------------------------------------
    // F# also has a "backwards pipe operator" but it's only recommended for
    // "undoing" one set of parens. If you guessed its symobol was `<|` you get
    // a golden star! If not, just think about the `<` and `>` is indicating the
    // direction that data flows

    let ``rewrite me using one back pipe`` =
      myAdd 1 <| myMult 2 6

    ensure (``rewrite me using one back pipe`` = 13)

    // Challenge 15-----------------------------------
    // Piping is pretty slick! Many functions in F# end up looking like
    let add1ThenMult5ThenAdd2ThenMult3 x =
      x
        |> myAdd 1
        |> myMult 5
        |> myAdd 2
        |> myMult 3

    // It reads pretty well, but we can take it a step further...
    // F# introduces a compose operator which uses the symbol `>>`

    // Compose says give me two functions, and, if one returns the input type
    // of the other, I'll glue them together and give you a new function!

    // So `add1ThenMult5ThenAdd2ThenMult3` can be rewritten without even talking about
    // the input value! Just by gluing functions together

    let doMathyStuffUsingCompose =
      myAdd 1
        >> myMult 5
        >> myAdd 2
        >> myMult 3

    // Notice it's basically the same thing except the `x` goes away and we replace
    // the `|` characters with `>`

    let refloogleThePreStewgledDrewgle x =
      x
        |> stewgle
        |> drewgle
        |> refloogle

    let refloogleUsingCompose =
      stewgle
        >> drewgle
        >> refloogle

    ensure (refloogleThePreStewgledDrewgle "space" = refloogleUsingCompose "space")

    // Challenge 16 -----------------------------------
    // "Backwards compose" also exists. Can you guess it's symbol?
    // (don't forget to wrap it in parens since it's a symbolic operator)

    let backComposeSymbol = (<<)

    ensure ((backComposeSymbol (myAdd 1) (myMult 4)) 2 = (2 * 4) + 1)

File: FpWorkshop/FpWorkshop/Step02_Functions.cs
using System;

namespace FpWorkshop
{
  public class Step02_Functions
  {
    public static T ITellYouWhatYouToldMe<T>(T x) => x;

    public static void Run()
    {
      // Challenge 1 --------------------------------------

      Help.Ensure(ITellYouWhatYouToldMe(42) == ReplaceMe.__<int>());

      // Challenge 2 --------------------------------------
      Func<int, int> echo = x => x;

      // I don't think there's a way to inline `x => x`...
      Help.Ensure(echo(42) == ReplaceMe.__<int>());

      // Challenge 3 --------------------------------------
      Func<int, int, int> myAdd = (a, b) => a + b;

      Help.Ensure(myAdd(2, 3) == ReplaceMe.__<int>());

      // Challenge 4 --------------------------------------
      Func<int, int> specialAdd = x => myAdd(4, x); // In C#, we need to re-wrap the function to build a new one

      Help.Ensure(specialAdd(2) == ReplaceMe.__<int>());

      // Challenge 4 (aside) ------------------------------
      // Functions that return functions are referred to as "curried" functions.
      // So, currying is possible in C# but not natural feeling
      // ♫ How many Funcs must a dev type out
      //   before they call it curried?
      //   the answer my friends is always one too many
      //   the answer is one too many ♫
      Func<int, Func<int, int>> myAddCurried = ReplaceMe.__<Func<int, Func<int, int>>>();

      var myNewSpecialAdd = myAddCurried(4);

      Help.Ensure(myNewSpecialAdd(2) == 6);

      // Challenge 5 --------------------------------------
      // Additional challenge -- can you define a version of `AlwaysReturnTheFirst` using a local variable?
      Help.Ensure(AlwaysReturnTheFirst<string, int>("foo")(42) == "foo");

      // Challenge 6 --------------------------------------
      // There's not really a great way to illustrate this in C#...

      // Challenge 7 --------------------------------------
      // We can define it in C# too! Though something's a little off...

      Let(5, a =>
      Let(6, b =>
      {
        Help.Ensure(a + b == 11);

        return "Why do we need a return here?";
      }));

      // Challenge 8 --------------------------------------
      // Compare the type we use here with the one F# infers for us
      Func<Func<int, int>, int, int, int> applyThenMultiply = (f, a, b) => f(a) * f(b);

      var addOne = myAddCurried(1);

      Help.Ensure(applyThenMultiply(addOne, 4, 5) == ReplaceMe.__<int>());

      // Challenge 9 --------------------------------------
      // This function's not built-in in C# but we've seen it before...
      Help.Ensure(applyThenMultiply(ReplaceMe.__<Func<int, int>>(), 7, 3) == (7 * 3));

      // Challenge 10 --------------------------------------
      Help.Ensure(applyThenMultiply(ReplaceMe.__<Func<int, Func<int, int>>>()(5), 7, 3) == (5 * 5));

      // No more challenges in this file
    }

    private static TResult Let<T, TResult>(T value, Func<T, TResult> body) =>
      ReplaceMe.__<TResult>();

    // Given this signature, how many possible implementations are there?
    private static Func<T2, T1> AlwaysReturnTheFirst<T1, T2>(T1 v) =>
      ReplaceMe.__<Func<T2, T1>>();
  }
}
